### ISSUE TITLE
MSSQL: executing the same query with cursor multiple time may result in duplicate rows

### ISSUE THREAD

**Issue by yosefdi** [dbeaver/dbeaver#4998]
When query is executed from SQL Editor window, it uses sp_cursoropen, sp_cursorfetch and sp_cursorclose stored procedures.
Scrolling the window results in issuing the same query again (different cursor), followed by sp_cursorfetch with ABSOLUTE fetch type.
E.g. the 1st cursor takes 400 rows. Than another cursor is opened with the same query and it is positioned to the row 400 via sp_cursorfetch.

**Please, notice that the order of rows in subsequent executions of the same query can be different**. If there is no ordering in the statement, the subsequent execution may produce at row 400 the row that in previous execution was at other location (e.g. 38). 

Therefore, the window will include some rows that are missing from the real result set and some are duplicate.

It is not a good practice to execute query multiple times, both from performance pov and from correctness pov. Also, better to avoid using cursors at all.

**Comment by serge-rider** [comment]
You can avoid duplicates by:
- Using transactional mode (disable auto-commit)
- Enable "Refresh data" option in connection or global settings:
![image](https://user-images.githubusercontent.com/6398845/51107951-95df8b00-1801-11e9-871b-9cac914beb61.png)

> It is not a good practice to execute query multiple times, both from performance pov and from correctness pov. Also, better to avoid using cursors at all.

Do you have any ideas how to do this better?  
Keeping open cursors "just in case" is much worse (and usually  not acceptable if you are working with some multi-user database).  
DBeaver doesn't open cursors explicitly. Some database drivers may do this internally. Usually this may be controlled by driver properties.

**Comment by yosefdi** [comment]
OK, got you.

Setting "Refresh data on next page reading" check-box alters behavior in a good way. Now sp_cursoropen is called once and followed by repeating sp_cursorfetch upon scrolling. Upon end of data set, sp_cursorclose is called. This is good. Since there is one result set, the data displayed is consistent with the data set.
The name and the description of the option is a bit cryptic to me and I don't that understand how it is describes the behavior change.

Looking on the driver properties, I see one property called 'selectMethod'. By default it is set to 'direct'. The other option is 'cursor'. So it is somewhat strange that even the property is set that way, cursors are still used.

I see other property there - 'responseBuffering'. I didn't notice any changes in behavior when I change this option.

I didn't find where can I disable auto-commit.

Bottom line for me: recommending to set "Refresh data on next page reading" solves the problem of consistency of the result set.

**Comment by serge-rider** [comment]
I think we'll make this option enabled by default in 5.3.3. There were quite many votes for this.

**Comment by mlt** [comment]
I suspect I have the same issue with PostgreSQL. It would be nice to have same default. I am on almost *master* there was a check mark by default.

**Comment by serge-rider** [comment]
It was made default in 6.x version

**Comment by yosefdi** [comment]
Great! Thanks.
